"IOException: Connection was closed" when retrying non-WebApplicationExceptions

### Describe the bug

Disclaimer: I asked it [here](https://github.com/quarkusio/quarkus/discussions/37313) already, because I am not sure if it is a bug. At least the behavior seems strange and not logical to me.

I have two local Quarkus applications, one - I will call it server - offers a simple REST endpoint, the other uses a RESTEasy Reactive RestClient to access this endpoint and has a retry configured in case of errors.

When I throw a WebApplicationException or a sub-type of it in the server endpoint, the client does a retry as expected.
When I throw any other Exception like e.g. a simple RuntimeException, the client receives a bunch of

```
2023-11-26 23:46:17,252 DEBUG [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0)
Restarting handler chain for exception exception: jakarta.ws.rs.ProcessingException:
java.io.IOException: Connection was closed
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:366)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:357)
(...)
2023-11-26 23:46:17,256 DEBUG [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-0) 
Attempting to handle unrecoverable exception: jakarta.ws.rs.ProcessingException:
java.io.IOException: Connection was closed
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:366)
	at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:357)
```

on the first retry attempt. No further retries are carried out then, it simply stops.

### Expected behavior

I would expect that the client does a retry in all HTTP 500 error cases because I cannot guarantee that only WebApplicationExceptions are thrown on the server side.

### Actual behavior

The client receives a HTTP 500 in both cases (the body differs though) but it is not able to retry on Exceptions which are not WebApplicationExceptions.

### How to Reproduce?

I created a reproducer [here](https://gitlab.pachamamita.de/pub/reproducer/restarting-handler-chain-connection-was-closed-issue). Follow the README.md to start the two applications.

### Output of `uname -a` or `ver`

Linux xxx 6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18 OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7) OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode, sharing)

### Quarkus version or git rev

3.5.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvnw 3.8.8

### Additional information

_No response_

Just want to add to this, I'm having similar issues with a container health check implementation utilizing a number of resteasy reactive rest clients. Despite completely avoiding any kind of retry logic, after a number of requests all subsequent requests start failing with the same IOException: Connection was closed error.

> : java.util.concurrent.CompletionException: java.io.IOException: Connection was closed  at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:79)  at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:65)  at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:46)  at org.company.common.rest.client.AbstractRestClientHealthCheck.healthCheck(AbstractRestClientHealthCheck.java:18)  at org.company.common.health.AbstractHealthCheck.call(AbstractHealthCheck.java:26)  at org.company.consumer.auth.health.DemographicsApiHealthCheck_ClientProxy.call(Unknown Source)  at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)  at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)  at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)  at io.smallrye.mutiny.operators.uni.UniOnFailureFlatMap.subscribe(UniOnFailureFlatMap.java:31)  at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)  at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)  at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)  at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)  at io.smallrye.mutiny.vertx.MutinyHelper.lambda$blockingExecutor$6(MutinyHelper.java:62)  at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:190)  at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:276)  at io.vertx.core.impl.ContextImpl.lambda$internalExecuteBlocking$2(ContextImpl.java:209)  at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)  at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)  at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1512)  at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)  at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)  at java.base@21.0.2/java.lang.Thread.runWith(Thread.java:1596)  at java.base@21.0.2/java.lang.Thread.run(Thread.java:1583)  at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:833)  at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211) Caused by: java.io.IOException: Connection was closed  at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:380)  at org.jboss.resteasy.reactive.client.handlers.ClientSendRequestHandler$4.handle(ClientSendRequestHandler.java:371)  at io.vertx.core.impl.future.FutureImpl$2.onFailure(FutureImpl.java:117)  at io.vertx.core.impl.future.FutureImpl$ListenerArray.onFailure(FutureImpl.java:303)  at io.vertx.core.impl.future.FutureBase.lambda$emitFailure$1(FutureBase.java:75)  at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:305)  at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:181)  at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:72)  at io.vertx.core.impl.future.FutureImpl.tryFail(FutureImpl.java:265)  at io.vertx.core.http.impl.HttpClientRequestBase.fail(HttpClientRequestBase.java:182)  at io.vertx.core.http.impl.HttpClientRequestBase.handleException(HttpClientRequestBase.java:177)  at io.vertx.core.http.impl.HttpClientRequestImpl.handleException(HttpClientRequestImpl.java:90)  at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleException(Http1xClientConnection.java:711)  at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleClosed(Http1xClientConnection.java:717)  at io.vertx.core.http.impl.Http1xClientConnection.lambda$handleClosed$12(Http1xClientConnection.java:1219)  at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:320)  at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:171)  at io.vertx.core.http.impl.Http1xClientConnection.handleClosed(Http1xClientConnection.java:1219)  at io.vertx.core.net.impl.VertxHandler.channelInactive(VertxHandler.java:143)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:303)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)  at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)  at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelInactive(CombinedChannelDuplexHandler.java:418)  at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:411)  at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:376)  at io.netty.handler.codec.http.HttpClientCodec$Decoder.channelInactive(HttpClientCodec.java:410)  at io.netty.channel.CombinedChannelDuplexHandler.channelInactive(CombinedChannelDuplexHandler.java:221)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:303)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)  at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)  at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:411)  at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:376)  at io.netty.handler.ssl.SslHandler.channelInactive(SslHandler.java:1174)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:303)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)  at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:274)  at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:301)  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:281)  at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)  at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:813)  at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)  at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)  at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)  at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)  at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)  ... 5 more

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java
@@ -640,7 +640,9 @@ public abstract class ResteasyReactiveRequestContext
 
     protected void endResponse() {
         if (serverResponse().headWritten()) {
-            serverRequest().closeConnection();
+            if (!serverResponse().closed()) {
+                serverRequest().closeConnection();
+            }
         } else {
             serverResponse().setStatusCode(500).end();
         }

--- a/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/VertxResteasyReactiveRequestContext.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/VertxResteasyReactiveRequestContext.java
@@ -406,7 +406,7 @@ public class VertxResteasyReactiveRequestContext extends ResteasyReactiveRequest
 
     @Override
     public boolean closed() {
-        return response.closed();
+        return response.ended() || response.closed();
     }
 
     @Override

